Multi-repo checkout does not work well with custom paths

## Agent Version and Platform
Agent version: '2.164.1' on macOS and Windows.

## Azure DevOps Type and Version
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3368518

## What's not working?
​I'm using the ​Multi repo checkout feature in a YAML pipeline with the following syntax:

```yaml
- checkout: foo
  clean: true  
  submodules: recursive  
  path: s/bar/external/foo
  persistCredentials: true
```

The post-job step which cleans up credentials in this repo is failing on both macOS and Windows. With diagnostic output enabled, you can see that the `targetPath` is incorrect. The clean up attempt tries to execute a git command in the default checkout location `s/foo` rather than the custom path I have specified `s/bar/external/foo`.

```
##[command]git config --unset-all http.https://github.com/org/foo.extraheader
##[debug]Starting process:
##[debug]  File name: '/usr/bin/git'
##[debug]  Arguments: 'config --unset-all http.https://github.com/org/foo.extraheader'
##[debug]  Working directory: '/Users/azdo/_work/5/s/foo'
...
##[debug]Exited process 26852 with exit code 2
##[debug]Finished process 26852 with exit code 2, and elapsed time 00:00:00.0027457.
```

This produces an error on Windows that causes the entire job to fail. On macOS this is only treated as a warning.

#### macOS:
```
##[warning]An unsuccessful attempt was made using git command line to remove "http.extraheader" from the git config. Attempting to modify the git config file directly to remove the credential.
```

#### Windows:
```
##[error]The directory name is invalid
##[debug]Processed: ##vso[task.logissue type=error;]The directory name is invalid
##[debug]Processed: ##vso[task.complete result=Failed;]
##[debug]   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at Microsoft.VisualStudio.Services.Agent.Util.ProcessInvoker.ExecuteAsync(String workingDirectory, String fileName, String arguments, IDictionary`2 environment, Boolean requireExitCodeZero, Encoding outputEncoding, Boolean killProcessOnCancel, InputQueue`1 redirectStandardIn, Boolean inheritConsoleHandler, Boolean keepStandardInOpen, Boolean highPriorityProcess, CancellationToken cancellationToken)
   at Agent.Plugins.Repository.GitCliManager.ExecuteGitCommandAsync(AgentTaskPluginExecutionContext context, String repoRoot, String command, String options, CancellationToken cancellationToken)
   at Agent.Plugins.Repository.GitCliManager.GitConfigUnset(AgentTaskPluginExecutionContext context, String repositoryPath, String configKey)
   at Agent.Plugins.Repository.GitSourceProvider.RemoveGitConfig(AgentTaskPluginExecutionContext executionContext, GitCliManager gitCommandManager, String targetPath, String configKey, String configValue)
   at Agent.Plugins.Repository.GitSourceProvider.PostJobCleanupAsync(AgentTaskPluginExecutionContext executionContext, RepositoryResource repository)
   at Agent.Plugins.Repository.CleanupTask.RunAsync(AgentTaskPluginExecutionContext executionContext, CancellationToken token)
   at Agent.PluginHost.Program.Main(String[] args)
```

I've got a pipeline (YAML with multiple checkout) failing since today. It might be related to the change.
What I've discovered so far: it looks like the directory **c:\agent_work\1\s** defined by predefined variables (https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml):
 - Build.Repository.LocalPath
 - Build.SourcesDirectory
 - System.DefaultWorkingDirectory
is not available (at least with deployment job) which gives an error: **##[error]Container path not found: 'C:\RubyAgent3\_work\42\s** for this trivial pipeline:

```
resources:
  repositories:
  - repository: infrastructure
    name: Infrastructure
    type: git

jobs:
- deployment: Test
  pool: 'Ruby Hosted Agents'
  environment: Test
  strategy:
    runOnce:
      deploy:
        steps:
        - checkout: infrastructure
          path: infrastructure
        - script: 'dir $(pipeline.workspace)'
```

> Current agent version: '2.163.1'

Regarding workaround: actually I found the problem with a different task: AzureResourceGroupDeployment@2 was failing with error: **##[error]The directory name is invalid**
It doesn't provide any information even with diagnostics enabled. So I started investigating and put some simple scripts to print out some info.

@jpricket I've extended the script in order to get more details and it looks like the issue is that the folder **Build.Repository.LocalPath** (e.g. c:\agent_work\1\s) is **moved** to a new name when you do checkout of some repo other than **self** to explicit path. Really moved itself, not just the content. Previously the content was moved, but directory `s` remained, so it worked. And this directory is used in many tasks and steps, so it's not working without it.

```
resources:
  repositories:
  - repository: infrastructure
    name: Infrastructure
    type: git

jobs:
- deployment: Test1
  pool: 'Windows Hosted Agents'
  environment: Test
  strategy:
    runOnce:
      deploy:
        steps:
        - checkout: infrastructure
          path: infrastructure
        - script: 'dir $(pipeline.workspace)' # !!!!!!! only this fails with ##[error]Container path not found: 'C:\WindowsAgent3_work\42\s
        - script: 'dir $(pipeline.workspace)' # this outputs directory list and directory s in not listed
          workingDirectory: $(pipeline.workspace)
          condition: always()
- deployment: Test2
  condition: always()
  pool: 'Windows Hosted Agents'
  environment: Test
  strategy:
    runOnce:
      deploy:
        steps:
        - checkout: infrastructure
        - script: 'dir $(pipeline.workspace)'
- deployment: Test3
  condition: always()
  pool: 'Windows Hosted Agents'
  environment: Test
  strategy:
    runOnce:
      deploy:
        steps:
        - checkout: self
        - script: 'dir $(pipeline.workspace)'
- deployment: Test4
  condition: always()
  pool: 'Windows Hosted Agents'
  environment: Test
  strategy:
    runOnce:
      deploy:
        steps:
        - script: 'dir $(pipeline.workspace)'
```

This is the output of the job Test1, please note that **subdirectory `s` is not there**
```
01/27/2020  09:18 PM    <DIR>          .
01/27/2020  09:18 PM    <DIR>          ..
01/27/2020  09:18 PM    <DIR>          a
01/27/2020  09:18 PM    <DIR>          b
01/27/2020  09:18 PM    <DIR>          infrastructure
01/27/2020  09:18 PM    <DIR>          TestResults
               0 File(s)              0 bytes
               6 Dir(s)  30,520,569,856 bytes free
```

All the other jobs passed and all listed subdirectory `s`

--- a/src/Agent.Worker/Build/TrackingManager.cs
+++ b/src/Agent.Worker/Build/TrackingManager.cs
@@ -368,6 +368,21 @@ namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
                 // The config is the new format.
                 Trace.Verbose("Parsing new tracking config format.");
                 result = JsonConvert.DeserializeObject<TrackingConfig>(content);
+                if (result != null)
+                {
+                    // if RepositoryTrackingInfo is empty, then we should create an entry so the rest
+                    // of the logic after this will act correctly
+                    if (result.RepositoryTrackingInfo.Count == 0)
+                    {
+                        result.RepositoryTrackingInfo.Add(new Build.RepositoryTrackingInfo
+                        {
+                            Identifier = RepositoryUtil.PrimaryRepositoryName,
+                            RepositoryType = result.RepositoryType,
+                            RepositoryUrl = result.RepositoryUrl,
+                            SourcesDirectory = result.SourcesDirectory,
+                        });
+                    }
+                }
             }
             else
             {

--- a/src/Test/L0/Worker/Build/TrackingManagerL0.cs
+++ b/src/Test/L0/Worker/Build/TrackingManagerL0.cs
@@ -142,11 +142,11 @@ namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build
                 string sourceFolder = Path.Combine(_workFolder, "b00335b6");
 
                 // It doesn't matter for this test whether the line endings are CRLF or just LF.
-                string Contents = @"{ 
-    ""system"" : ""build"", 
-    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"", 
-    ""definitionId"" = ""7"", 
-    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"", 
+                string Contents = @"{
+    ""system"" : ""build"",
+    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"",
+    ""definitionId"" = ""7"",
+    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"",
     ""sourceFolder"" = """ + sourceFolder + @""",
     ""hashKey"" = ""b00335b6923adfa64f46f3abb7da1cdc0d9bae6c""
 }";
@@ -177,11 +177,11 @@ namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build
                 string sourceFolder = Path.Combine(_workFolder, "b00335b6");
 
                 // It doesn't matter for this test whether the line endings are CRLF or just LF.
-                string contents = @"{ 
-    ""system"" : ""build"", 
-    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"", 
-    ""definitionId"" = ""7"", 
-    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"", 
+                string contents = @"{
+    ""system"" : ""build"",
+    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"",
+    ""definitionId"" = ""7"",
+    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"",
     ""sourceFolder"" = """ + sourceFolder + @""",
     ""hashKey"" = """"
 }";
@@ -208,11 +208,11 @@ namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build
                 string sourceFolder = Path.Combine(_workFolder, "b00335b6");
 
                 // It doesn't matter for this test whether the line endings are CRLF or just LF.
-                string contents = @"{ 
-    ""system"" : ""build"", 
-    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"", 
-    ""definitionId"" = ""7"", 
-    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"", 
+                string contents = @"{
+    ""system"" : ""build"",
+    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"",
+    ""definitionId"" = ""7"",
+    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"",
     ""sourceFolder"" = """ + sourceFolder + @""",
     ""hashKey"" = ""b00335b6923adfa64f46f3abb7da1cdc0d9bae6c""
 }";
@@ -466,6 +466,14 @@ namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build
   ""agent_builddirectory"": ""b00335b6"",
   ""collectionUrl"": ""http://contoso:8080/tfs/DefaultCollection/"",
   ""definitionName"": ""M87_PrintEnvVars"",
+  ""repositoryTrackingInfo"": [
+    {
+      ""identifier"": ""self"",
+      ""repositoryType"": ""tfsgit"",
+      ""repositoryUrl"": ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"",
+      ""sourcesDirectory"": ""b00335b6\\gitTest""
+    }
+  ],
   ""fileFormatVersion"": 3,
   ""lastRunOn"": ""09/16/2015 23:56:46 -04:00"",
   ""repositoryType"": ""tfsgit"",
@@ -520,10 +528,10 @@ namespace Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build
 
                 // It doesn't matter for this test whether the line endings are CRLF or just LF.
                 string trackingContents = @"{
-    ""system"" : ""build"", 
-    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"", 
-    ""definitionId"" = ""7"", 
-    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"", 
+    ""system"" : ""build"",
+    ""collectionId"" = ""7aee6dde-6381-4098-93e7-50a8264cf066"",
+    ""definitionId"" = ""7"",
+    ""repositoryUrl"" = ""http://contoso:8080/tfs/DefaultCollection/_git/gitTest"",
     ""sourceFolder"" = """ + sourceFolder + @""",
     ""hashKey"" = ""b00335b6923adfa64f46f3abb7da1cdc0d9bae6c""
 }";

